### PR TITLE
[runtime env] Create virtualenv for pip runtime env.

### PR DESCRIPTION
**Pull request by fyrestone**
## Why are these changes needed?

Runtime env improvements:
- [X] Create virtualenv for pip runtime env.
- [X] Raise the original traceback from runtime env agent.
- [X] Runtime env context compatible with PyCharm monkey patch.
- [X] Test case for pip runtime env checking ray.

Other fixes:
- [X] `setup_worker.py` and `ray/util/client/server/server.py` use ray logging config.
- [X] `test_utils.py::chdir` restores cwd when got error in yield.
- [X] Skip empty out lines of `exec_cmd_stream_to_logger`.

## Checks

- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [X] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [X] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by SongGuyang — APPROVED** [review]
LGTM. @edoakes Can you take a look?

File: python/setup.py
Comment (edoakes) [inline python/setup.py:213]: @rkooo567 is currently making it so runtime_env is supported in the basic installation rather than `ray[default]`, so we'd want to put it there instead.

**Comment by edoakes** [comment]
This seems like a good improvement @fyrestone, I'm a bit concerned that we're going to bloat the "minimal" ray install again though. Is it possible that we make this an optional dependency? i.e., if you try to use the `pip` field and don't have virtualenv installed, it throws an error?

File: python/ray/_private/runtime_env/pip.py
Comment (fyrestone) [reply python/ray/_private/runtime_env/pip.py:50]: `PipManager` setup and delete uri, it's the interface of the pip. While the `PipProcessor` only process the pip runtime env, it's the implementation of the pip. The processor class may be reused in other manager, e.g. `ContainerManager`, then user can setup pip based on a custom docker image.

File: python/ray/_private/runtime_env/pip.py
Comment (fyrestone) [reply python/ray/_private/runtime_env/pip.py:119]: FIxed. Thanks.

Known issue: Raising an exception here has potential compatible issues, e.g. If the runtime env agent is running inside a virtualenv, then the pip runtime env creation will fail.

File: python/ray/_private/runtime_env/pip.py
Comment (fyrestone) [reply python/ray/_private/runtime_env/pip.py:154]: Fixed. Thanks.

File: python/setup.py
Comment (fyrestone) [reply python/setup.py]: I moved `virtualenv` to the `setup_spec.install_requires` in setup.py

**Comment by fyrestone** [comment]
> This seems like a good improvement @fyrestone, I'm a bit concerned that we're going to bloat the "minimal" ray install again though. Is it possible that we make this an optional dependency? i.e., if you try to use the `pip` field and don't have virtualenv installed, it throws an error?

Sure. I will change the virtualenv as an optional dependency.

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Comment by architkulkarni** [comment]
cc @rkooo567 there are error handling and logging improvements in this PR which may be related to what you're working on!

File: python/ray/_private/runtime_env/pip.py
Comment (architkulkarni) [inline python/ray/_private/runtime_env/pip.py:237]: Why not use the `f`-string?

**Review by architkulkarni — APPROVED** [review]
Looks good to me!  Thanks for the logging fixes.  I just had a few minor questions.

File: python/ray/_private/runtime_env/pip.py
Comment (SongGuyang) [reply python/ray/_private/runtime_env/pip.py:125]: @architkulkarni I'm also worried about this. @fyrestone Can we make this "TODO" to be a P0 issue? We should support creating virtualenv from virtualenv as soon as possible.

**Comment by fyrestone** [comment]
> ## ‼️ ACTION REQUIRED ‼️
> We've switched our code formatter from YAPF to Black (see #21311).
> 
> To prevent issues with merging your code, here's what you'll need to do:
> 
> 1. Install Black
> 
> ```
> pip install -I black==21.12b0
> ```
> 
> 2. Format changed files with Black
> 
> ```
> curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
> chmod +x ./format-changed.sh
> ./format-changed.sh
> rm format-changed.sh
> ```
> 
> 3. Commit your changes.
> 
> ```
> git add --all
> git commit -m "Format Python code with Black"
> ```
> 
> 4. Merge master into your branch.
> 
> ```
> git pull upstream master
> ```
> 
> 5. Resolve merge conflicts (if necessary).
> 
> After running these steps, you'll have the updated `format.sh`.

Thanks. I will format my PR by black.

File: python/ray/_private/runtime_env/pip.py
Comment (fyrestone) [reply python/ray/_private/runtime_env/pip.py]: When this error message appears, it should come from `runtime_env` pip. I will add the message to make it more clear.

File: python/ray/_private/runtime_env/pip.py
Comment (fyrestone) [reply python/ray/_private/runtime_env/pip.py:241]: logger.exception prints the traceback with the exception string.

File: python/ray/_private/runtime_env/pip.py
Comment (fyrestone) [reply python/ray/_private/runtime_env/pip.py:267]: logger filters the log level first, then eval the var. Use % formatter can reduce the eval cost when the logging level mismatch. If we use f-string here, the `{pip_env_path}` will always be evaluated even we set the logging level to ERROR.

File: python/ray/_private/runtime_env/pip.py
Comment (fyrestone) [reply python/ray/_private/runtime_env/pip.py]: Thanks. Comments and test case are added.

**Comment by edoakes** [comment]
@architkulkarni can you verify that this doesn't reinstate the bug you fixed recently with non-package entries in the pip list?

File: python/ray/_private/runtime_env/pip.py
Comment (architkulkarni) [reply python/ray/_private/runtime_env/pip.py]: Ah nice, I didn't know that!

**Comment by architkulkarni** [comment]
> @architkulkarni can you verify that this doesn't reinstate the bug you fixed recently with non-package entries in the pip list?

Looks good, it looks like this PR independently fixes it the same way, by dumping all the lines to a requirements.txt file:
https://github.com/ray-project/ray/pull/21801/files#diff-c2ae3a97f1b2217b319f7ab97b678fe70d3319eb690405bb263b3b925a501392R196-R199
Actually this PR also entirely bypasses the part where we scan the pip list for Ray packages, which is the part that introduced the bug. 

For reference, the fix we're referring to is https://github.com/ray-project/ray/pull/22065.

**Comment by architkulkarni** [comment]
Looks like `//python/ray/tests:test_runtime_env_conda_and_pip` timed out, @fyrestone can you try splitting it into two files?

**Comment by fyrestone** [comment]
> Looks like `//python/ray/tests:test_runtime_env_conda_and_pip` timed out, @fyrestone can you try splitting it into two files?

I moved new test cases to `test_runtime_env_conda_and_pip_4.py`. I think we can remove the rewrite logic in the future, because user can install the optional dependencies by `ray[xxx]`  in runtime env.

File: python/ray/_private/runtime_env/pip.py
Comment (fyrestone) [reply python/ray/_private/runtime_env/pip.py:132]: > @architkulkarni I'm also worried about this. @fyrestone Can we make this "TODO" to be a P0 issue? We should support creating virtualenv from virtualenv as soon as possible.

Sure, I will create a PR to support creating virtualenv from virtualenv after this PR is merged.

**Comment by fyrestone** [comment]
> Looks like it's still timing out unfortunately, I wonder if it makes sense to move more tests to the new file?
> 
> The new test failed here https://buildkite.com/ray-project/ray-builders-pr/builds/24118#d709a7a9-7c92-4cdd-8d0d-00126a3f4e60/6-7713

What's the difference of `test_actor_level_gc` in `test_runtime_env_conda_and_pip.py` and `test_runtime_env_conda_and_pip_3.py`?

**Comment by fyrestone** [comment]
> Looks like it's still timing out unfortunately, I wonder if it makes sense to move more tests to the new file?
> 
> The new test failed here https://buildkite.com/ray-project/ray-builders-pr/builds/24118#d709a7a9-7c92-4cdd-8d0d-00126a3f4e60/6-7713

I think they are same: 
https://github.com/ray-project/ray/blob/master/python/ray/tests/test_runtime_env_conda_and_pip.py#L216
https://github.com/ray-project/ray/blob/master/python/ray/tests/test_runtime_env_conda_and_pip_3.py#L28

I will remove the `test_actor_level_gc` from `test_runtime_env_conda_and_pip.py`.

**Review by raulchen — APPROVED** [review]

**Comment by fyrestone** [comment]
@edoakes hi, this PR needs code owner reviewed. Can you merge it?

**Review by edoakes — APPROVED** [review]

**Comment by Mazyod** [comment]
> This seems like a good improvement @fyrestone, I'm a bit concerned that we're going to bloat the "minimal" ray install again though. Is it possible that we make this an optional dependency? i.e., if you try to use the `pip` field and don't have virtualenv installed, it throws an error?

When we updated to ray 1.12.0 with poetry, it pulled virtualenv as a dependency anyway. We are also concerned about the bloating of the minimal ray installation. Not sure though if this is expected when using poetry or not.

**Comment by SongGuyang** [comment]
> > This seems like a good improvement @fyrestone, I'm a bit concerned that we're going to bloat the "minimal" ray install again though. Is it possible that we make this an optional dependency? i.e., if you try to use the `pip` field and don't have virtualenv installed, it throws an error?
> 
> When we updated to ray 1.12.0 with poetry, it pulled virtualenv as a dependency anyway. We are also concerned about the bloating of the minimal ray installation. Not sure though if this is expected when using poetry or not.

Hi @Mazyod, can you create a new issue about this? It is more easy to discuss and trice, instead of this merged PR.

**Pull request merged**
